scmailclient/scmail.py

```diff
@@ -0,0 +1,493 @@
+import click
+import pprint
+import re
+from pathlib import Path
+from os import environ
+import getpass
+import logging
+import time
+import requests
+
+
+from gpgopt import GpgOpt
+
+
+# Default to localhost if url is not given
+SECUREMAILBOX_URL = environ.get(
+    "SECUREMAILBOX_URL", "http://securemailbox.duckdns.org:8082/"
+)
+
+# Logging Level
+LOGGING_LEVEL = environ.get("LOGGING_LEVEL", logging.DEBUG)
+
+
+def check_fingerprint(ctx, param, value):
+    # in normal case or it is optional.
+    if len(value) == 40 or not value:
+        return value
+    logging.error("The length of fingerprint is not a valid length.")
+    ctx.exit(1)
+
+
+@click.group(chain=True)
+@click.pass_context
+def client(ctx):
+    """Initial client."""
+    # Bacis config of logging
+    Path("log").mkdir(exist_ok=True)
+    times = time.strftime("%Y%m%d%H%M%S", time.localtime())
+    logging.basicConfig(
+        format="%(asctime)s %(name)s %(levelname)s %(message)s",
+        level=LOGGING_LEVEL,
+        handlers=[
+            logging.FileHandler(filename=Path("log", times)),
+            logging.StreamHandler(),
+        ],
+    )
+    ctx.logger = logging.getLogger(__name__)
+
+    # Set config file path.
+    ctx.config = Path("scmail.conf")
+
+    # If the config not exist, set user information. If not, run initial gpg.
+    gnupghome = (
+        load_config()
+        if ctx.config.exists()
+        else ctx.invoke(
+            set_user_info,
+            gnupghome=click.prompt("Enter your gnupg home dir:", default="gnupgkeys"),
+        )
+    )
+    ctx.logger.debug(f"gnupg home is {gnupghome}")
+
+    ctx.gpg = GpgOpt(mygnupghome=gnupghome)
+
+    # Initial pretty printer
+    ctx.pp = pprint.PrettyPrinter(indent=2)
+
+    logging.info("Initialize client successful.")
+
+
+@click.pass_context
+def load_config(ctx):
+    """Load config file."""
+    f = ctx.config.open("r").read()
```

Comment:
This file doesn't look like it's being closed after it's read from. I would recommend using `with ctx.config.open("r) as f: ...` instead as it will automatically close the file for you.

Reply:
Thanks, I use `with ... as f:`